docs/src/proposals/partitioned-inflationary-rewards-distribution.md

```diff
@@ -0,0 +1,110 @@
+---
+title: Partitioned Inflationary Rewards Distribution
+---
+
+## Problem
+
+With the increase of number of stake accounts, computing and redeeming the stake
+rewards at the start block of the epoch boundary becomes very expensive.
+Currently, with 550K stake accounts, the stake reward time has already taken
+more than 10 seconds. This prolonged computation slows down the network, and can
+cause large number of forks at the epoch boundary, which makes the matter even
+worse.
+
+## Proposed Solutions
+
+Instead of computing and reward stake accounts at epoch boundary, we will
+decouple reward computation and reward credit into two phases. At the start of
+the new epoch, a background thread is forked to compute the rewards. This is the
```

Comment:
won't the thread already be created and re-used? if it's started at the epoch start but then not used until N- epoch end, what does it do until then, it just waits?

Reply:
This question and the above question about the forking across the epoch boundaries, make me think that we should have a long-running service to calculate the rewards. 
For any forks that cross the epoch boundaries, we will need to calculate the rewards. We could create a channel to send and receive calculation requests between the blocks and the service. 
To avoid repeated computation with the same request, we can generate a signature for each request as epoch_number + hash(all_stake_accounts). 
For the example above, if stake_accounts didn't change between slot N-1 and slot N, we can reuse the computation for slot N+1 and N+2.

Reply:
Updated the proposal to address the comments.